src/rust/engine/fs/src/store.rs

```diff
@@ -212,6 +212,89 @@ impl Store {
       .to_boxed()
   }
 
+  ///
+  /// Ensures that the remote ByteStore has a copy of each passed Fingerprint, including any files
+  /// contained in any Directories in the list.
+  ///
+  /// At some point in the future we may want to make a call to
+  /// ContentAddressableStorage.FindMissingBlobs to avoid uploading things which are already present
+  /// remotely, but this optimization has not yet been made.
+  ///
+  pub fn ensure_remote_has_recursive(
+    &self,
+    fingerprints: Vec<Fingerprint>,
+  ) -> BoxFuture<(), String> {
+    let remote = match self.remote {
+      Some(ref remote) => remote,
+      None => {
+        return future::err(format!("Cannot ensure remote has blobs without a remote")).to_boxed()
```

Comment:
Rather than an error, should this method be rephrased such that it is safe to call regardless of whether a remote is configured? Maybe `ensure_synced_recursive` or something? Otherwise it seems like the caller is guaranteed to need a conditional that checks whether there is a remote before calling this. Maybe that conditional will already exist though.

Reply:
Yeah, I considered this, but I think the only real caller is going to be in the remote process executor, and it'd be better for that to hard-fail on upload attempt. The alternatives are either:
1. Have the `Store` expose whether it has a remote, and add a conditional there, or
2. Have the remote executor start counting missing digest errors per-digest so that it can detect that it keeps having the same files missing from the remote, and guess that this is the problem.

(`fs_util` will also use this, but it can just keep track of whether it has a remote really easily :))